server/errors/errors.go

```diff
@@ -0,0 +1,174 @@
+package errors
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	redigo "github.com/gomodule/redigo/redis"
+	"github.com/mna/redisc"
+	"github.com/rotisserie/eris"
+)
+
+var errCh chan error
+var running int32
+
+type errorHandler struct {
+	pool   fleet.RedisPool
+	logger kitlog.Logger
+}
+
+type ErrorFlusher interface {
+	Flush() ([]string, error)
+}
+
+func NewHandler(ctx context.Context, pool fleet.RedisPool, logger kitlog.Logger) ErrorFlusher {
+	errCh = make(chan error)
+
+	e := &errorHandler{
+		pool:   pool,
+		logger: logger,
+	}
+	go e.handleErrors(ctx)
+	return e
+}
+
+func (e *errorHandler) Flush() ([]string, error) {
+	errorKeys, err := redis.ScanKeys(e.pool, "error:*")
+	if err != nil {
+		return nil, err
+	}
+	keysBySlot := redis.SplitRedisKeysBySlot(e.pool, errorKeys...)
+	var errors []string
+	for _, qkeys := range keysBySlot {
+		gotErrors, err := e.collectBatchErrors(qkeys)
+		errors = append(errors, gotErrors...)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return errors, nil
+}
+
+func (e *errorHandler) collectBatchErrors(errorKeys []string) ([]string, error) {
+	var errors []string
+
+	conn := e.pool.Get()
+	defer conn.Close()
+
+	if rc, err := redisc.RetryConn(conn, 3, 100*time.Millisecond); err == nil {
+		conn = rc
+	}
+	for _, key := range errorKeys {
+		errorJson, err := redigo.String(conn.Do("GET", key))
+		if err != nil {
+			return nil, err
+		}
+		errors = append(errors, errorJson)
+		_, err = conn.Do("DEL", key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return errors, nil
+}
+
+func sha256b64(s string) string {
+	src := sha256.Sum256([]byte(s))
+	return base64.URLEncoding.EncodeToString(src[:])
+}
+
+func hashErrorLocation(err error) string {
+	unpackedErr := eris.Unpack(err)
+	if unpackedErr.ErrExternal != nil {
+		return sha256b64(unpackedErr.ErrExternal.Error())
+	}
+
+	if len(unpackedErr.ErrRoot.Stack) == 0 {
+		return sha256b64(unpackedErr.ErrRoot.Msg)
+	}
+
+	lastFrame := unpackedErr.ErrRoot.Stack[0]
+	return sha256b64(fmt.Sprintf("%s:%d", lastFrame.File, lastFrame.Line))
```

Comment:
So if there's a stack trace, it only hashes the location? I'm kind of wondering if that is unique enough - i.e. something like writing to a file descriptor can fail in various ways, I'm not sure if that would necessarily have a different location in the Go stack trace? Thinking out loud a bit, but that's probably fine, I guess it would have to have a different location to create the Go error.

Reply:
I'm trying to mimic what Sentry does, sort of. It hashes where the error originates. If we hash many more things, then the chances of it creating a ton of redis entries is higher. But there's so much more room for improvement anyway, just had to make a call to start.

I suppose we could add the type of the error to the hash.

Reply:
Yeah just to be safe I like the idea of adding the type.